lib/logstash/filters/geoip.rb

```diff
@@ -82,139 +86,121 @@ class LogStash::Filters::GeoIP < LogStash::Filters::Base
   # of the same geoip_type share the same cache. The last declared cache size will 'win'. The reason for this is that there would be no benefit
   # to having multiple caches for different instances at different points in the pipeline, that would just increase the
   # number of cache misses and waste memory.
-  config :lru_cache_size, :validate => :number, :default => 1000
+  config :cache_size, :validate => :number, :default => 1000
 
   public
   def register
-    require "geoip"
-
     if @database.nil?
-      @database = ::Dir.glob(::File.join(::File.expand_path("../../../vendor/", ::File.dirname(__FILE__)),"GeoLiteCity*.dat")).first
-      if !File.exists?(@database)
-        raise "You must specify 'database => ...' in your geoip filter (I looked for '#{@database}'"
+      @database = ::Dir.glob(::File.join(::File.expand_path("../../../vendor/", ::File.dirname(__FILE__)),"GeoLite2-City.mmdb")).first
+
+      if @database.nil? || !File.exists?(@database)
+        raise "You must specify 'database => ...' in your geoip filter (I looked for '#{@database}')"
       end
     end
-    @logger.info("Using geoip database", :path => @database)
-    # For the purpose of initializing this filter, geoip is initialized here but
-    # not set as a global. The geoip module imposes a mutex, so the filter needs
-    # to re-initialize this later in the filter() thread, and save that access
-    # as a thread-local variable.
-    geoip_initialize = ::GeoIP.new(@database)
-
-    @geoip_type = case geoip_initialize.database_type
-    when GeoIP::GEOIP_CITY_EDITION_REV0, GeoIP::GEOIP_CITY_EDITION_REV1
-      :city
-    when GeoIP::GEOIP_COUNTRY_EDITION
-      :country
-    when GeoIP::GEOIP_ASNUM_EDITION
-      :asn
-    when GeoIP::GEOIP_ISP_EDITION, GeoIP::GEOIP_ORG_EDITION
-      :isp
-    else
-      raise RuntimeException.new "This GeoIP database is not currently supported"
-    end
 
-    @threadkey = "geoip-#{self.object_id}"
+    @logger.info("Using geoip database", :path => @database)
 
-    # This is wrapped in a mutex to make sure the initialization behavior of LOOKUP_CACHES (see def above) doesn't create a dupe
-    LOOKUP_CACHE_INIT_MUTEX.synchronize do
-      self.lookup_cache = LOOKUP_CACHES[@geoip_type] ||= LruRedux::ThreadSafeCache.new(1000)
+    db_file = JavaIO::File.new(@database)
+    begin
+      @parser = DatabaseReader::Builder.new(db_file).withCache(CHMCache.new(@cache_size)).build();
+    rescue Java::ComMaxmindDb::InvalidDatabaseException => e
+      # do nothing
```

Comment:
@talevy When I add `logger.error` to line 107, it cannot pass spec test, how to solve this ?

The code I added is 
```
@logger.error("Is the Geoip2 MMDB database invalid or corrupted ?", :exception => e, :field => @source)
```

The Error bundle exec rspec output:
```
Failures:
  1) LogStash::Filters::GeoIP an invalid database should return the correct sourcefield in the logging message "{"ip":"8.8.8.8"}" when processed
     Failure/Error: expect(LogStash::Filters::GeoIP.logger).to receive(:error).with(anything, include(:field => "ip"))
       (#<Cabin::Channel:0x7d2b82ef>).error(anything, include {:field => "ip"})
           expected: 1 time with arguments: (anything, include {:field => "ip"})
           received: 2 times with arguments: (#<RSpec::Mocks::ArgumentMatchers::AnyArgMatcher:0x5af6ae34>, #<RSpec::Matchers::BuiltIn::Include:0x79270693 @actual={:exception=
>#<NoMethodError: undefined method `city' for nil:NilClass>, :field=>"ip", :event=>#<LogStash::Event:0x240c243e @metadata={}, @accessors=#<LogStash::Util::Accessors:0x191f74e
f @store={"ip"=>"8.8.8.8", "@version"=>"1", "@timestamp"=>"2016-03-24T02:19:15.058Z", "geoip"=>{}}, @lut={"ip"=>[{"ip"=>"8.8.8.8", "@version"=>"1", "@timestamp"=>"2016-03-24T
02:19:15.058Z", "geoip"=>{}}, "ip"], "geoip"=>[{"ip"=>"8.8.8.8", "@version"=>"1", "@timestamp"=>"2016-03-24T02:19:15.058Z", "geoip"=>{}}, "geoip"]}>, @data={"ip"=>"8.8.8.8",
"@version"=>"1", "@timestamp"=>"2016-03-24T02:19:15.058Z", "geoip"=>{}}, @metadata_accessors=#<LogStash::Util::Accessors:0x29b1b28c @store={}, @lut={}>, @cancelled=false>}, @
expected=[{:field=>"ip"}]>)
     # ./spec/filters/geoip_spec.rb:212:in `(root)'
     # /home/nitrous/.rvm/gems/jruby-1.7.19/gems/rspec-wait-0.0.8/lib/rspec/wait.rb:46:in `(root)'
Finished in 1.74 seconds (files took 4.64 seconds to load)
17 examples, 1 failure
Failed examples:
rspec ./spec/filters/geoip_spec.rb:211 # LogStash::Filters::GeoIP an invalid database should return the correct sourcefield in the logging message "{"ip":"8.8.8.8"}" when pro
cessed
```